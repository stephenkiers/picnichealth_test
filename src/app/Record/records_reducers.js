import { combineReducers } from 'redux-immutable'
import record from './record_reducers';
import { Map, Set } from 'immutable'

export const record_constants = {
    RECORD_DELETE: 'RECORD_DELETE',
    RECORD_UPDATE: 'RECORD_UPDATE',
    RECORD_UPDATE_STATE: 'RECORD_UPDATE_STATE',
};


const byId = function(state = Map(), action) {
    switch (action.type) {
        case record_constants.RECORD_UPDATE:
        case record_constants.RECORD_UPDATE_STATE:
            return state.update(action.id, r => record(r, action))
        case record_constants.RECORD_DELETE:
            return state.delete(action.id)
        default:
            return state;
    }
};

const records = combineReducers({
    byId
});

export default records;

export const getRecords = state =>
    state.get('byId').valueSeq().toSet();

export const getRecord = (state, id) =>
    state.getIn(['byId', id]);