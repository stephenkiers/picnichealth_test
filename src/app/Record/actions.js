import {record_constants} from './records_reducers'
import {reduxAppStateSetModalId} from "../AppState/actions";
import faker from 'faker';
import {generateRandomUUID} from "../utils";

export const reduxRecordUpdate = (id, first_name, last_name, phone_number, created_at) => ({
    type: record_constants.RECORD_UPDATE,
    id,
    first_name,
    last_name,
    phone_number,
    created_at,
});
export const reduxRecordDelete = id => ({
    type: record_constants.RECORD_DELETE,
    id
})
export const reduxRecordUpdateState = (id, state) => ({
    type: record_constants.RECORD_UPDATE_STATE,
    id,
    state,
});

export const apiRecordGetLatest = () => {
    return dispatch => {
        const fake_fetch = new Promise((resolve, reject) => {
            setTimeout(function(){
                let testData = []
                Array.from(new Array(10)).forEach(() => {
                    testData.push({
                        id: generateRandomUUID(),
                        first_name: faker.name.firstName(),
                        last_name: faker.name.lastName(),
                        phone_number: `${Math.floor(1000000000 + Math.random() * 9000000000)}`,
                        created_at: (new Date).getTime() - Math.floor(Math.random() * 1000)
                    })
                })
                resolve(testData);
            }, 500);
        });
        return fake_fetch.then(res => {
            // GET 'v1/records'

            // this is not an ideal seed file, but I don't think it needs to be
            // optimized for this test, as a backend would really be used
            res.forEach(result => {
                dispatch(reduxRecordUpdate(
                    result.id,
                    result.first_name,
                    result.last_name,
                    result.phone_number,
                    result.created_at
                ));
            })
        }).catch(err => {
            //handleErr
        })
    }
}


export const apiRecordUpdate = (id, first_name, last_name, phone_number, created_at) => {
    return dispatch => {
        dispatch(reduxRecordUpdateState(id, 'saving'));
        created_at = created_at || (new Date).getTime();
        const fake_fetch = new Promise((resolve, reject) => {
            //This should be a fetch promise to send attributes to backend.
            // In practice, there would be a catch when/if there are errors
            // POST `v1/records/${id}`
            setTimeout(function() {
                resolve()
            }, 2000)
        });
        return fake_fetch.then(res => {
            dispatch(reduxRecordUpdate(id, first_name, last_name, phone_number, created_at));
            dispatch(reduxRecordUpdateState(id, 'idle'));
            dispatch(reduxAppStateSetModalId(undefined));
        })
    }
}
export const apiRecordDelete = id => {
    return dispatch => {
        dispatch(reduxRecordUpdateState(id, 'deleting'))
        const fake_fetch = new Promise((resolve, reject) => {
            //This should be a fetch promise to send attributes to backend.
            // In practice, there would be a catch when/if there are errors
            // DELETE `v1/records/${id}`
            setTimeout(function() {
                resolve()
            }, 1000)
        });
        return fake_fetch.then(res => {
            dispatch(reduxRecordDelete(id));
            dispatch(reduxAppStateSetModalId(undefined));
        });
    }
}