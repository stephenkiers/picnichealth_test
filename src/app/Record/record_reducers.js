import { record_constants } from './records_reducers'
import { Map } from 'immutable'

export const default_record = Map({
    id: "",
    first_name: "",
    last_name: "",
    phone_number: "",
    created_at: undefined,
})

const city = (state = Map({state: 'idle'}), action) => {
    switch (action.type) {
        case record_constants.RECORD_UPDATE:
            return state.merge(Map({
                id: action.id,
                first_name: action.first_name,
                last_name: action.last_name,
                phone_number: action.phone_number,
                created_at: action.created_at
            }));
        case record_constants.RECORD_UPDATE_STATE:
            return state.merge(Map({
                id: action.id,
                state: action.state,
            }));

        default:
            return state;
    }
}
export default city