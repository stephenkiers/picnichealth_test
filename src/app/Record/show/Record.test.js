import React from 'react';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import Record from './Record';

test('Record displays correct', () => {
    const renderer = new ReactShallowRenderer();
    renderer.render(
        <Record
            id="1234567890"
            first_name="FirstName"
            last_name="LastName"
            phone_number="5555555555"
            onUpdateClick={() => {}}
            onDelete={() => {}}
        />
    );
    const tree = renderer.getRenderOutput()
    expect(tree).toMatchSnapshot();
});