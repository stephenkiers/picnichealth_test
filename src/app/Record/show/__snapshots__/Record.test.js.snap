// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Record displays correct 1`] = `
<div
  className="record clearfix"
>
  <div
    className="pull-left"
  >
    <strong
      className="last-name"
    >
      LastName
      ,
       
    </strong>
    <span
      className="first-name"
    >
      FirstName
    </span>
    <a
      className="phone-number"
      href="tel:5555555555"
    >
      555-555-5555
    </a>
  </div>
  <div
    className="pull-right"
  >
    <a
      className="btn"
      href="javascript:void(0);"
      onClick={[Function]}
    >
      Update
    </a>
    <a
      className="btn btn-danger"
      href="javascript:void(0);"
      onClick={[Function]}
    >
      Delete
    </a>
  </div>
</div>
`;
